src/socket.rs

```diff
@@ -1,56 +1,209 @@
-use mio::unix::EventedFd;
-use mio::{Evented, Poll, PollOpt, Ready, Token};
-use zmq::Socket;
-
 use std::io;
 
-//Wrapper type to allow us to impl Evented & PollEvented
-pub struct MioSocket {
-    pub io: Socket,
-}
+use mio::{unix::EventedFd, Evented, Poll, PollOpt, Ready, Token};
 
-impl MioSocket {
-    pub fn new(socket: Socket) -> Self {
-        Self { io: socket }
-    }
-}
+use crate::Result;
 
-//Convert to/from MioSocket
-impl From<Socket> for MioSocket {
-    fn from(socket: Socket) -> Self {
-        Self { io: socket }
-    }
+pub(crate) struct SocketWrapper {
+    pub(crate) socket: zmq::Socket,
 }
 
-//Convert to/from Socket
-impl From<MioSocket> for Socket {
-    fn from(socket: MioSocket) -> Self {
-        socket.io
+impl SocketWrapper {
+    pub fn new(socket: zmq::Socket) -> Self {
+        Self { socket }
     }
 }
 
-impl Evented for MioSocket {
+impl Evented for SocketWrapper {
     fn register(
         &self,
         poll: &Poll,
         token: Token,
         interest: Ready,
         opts: PollOpt,
     ) -> io::Result<()> {
-        EventedFd(&self.io.get_fd()?).register(poll, token, interest, opts)
+        EventedFd(&self.socket.get_fd()?).register(poll, token, interest, opts)
     }
-
     fn reregister(
         &self,
         poll: &Poll,
         token: Token,
         interest: Ready,
         opts: PollOpt,
     ) -> io::Result<()> {
-        EventedFd(&self.io.get_fd()?).reregister(poll, token, interest, opts)
+        EventedFd(&self.socket.get_fd()?).reregister(poll, token, interest, opts)
     }
-
     fn deregister(&self, poll: &Poll) -> io::Result<()> {
-        EventedFd(&self.io.get_fd()?).deregister(poll)
+        EventedFd(&self.socket.get_fd()?).deregister(poll)
     }
 }
+
+pub trait AsZmqSocket {
+    #[inline]
+    fn get_socket(&self) -> &zmq::Socket;
+}
+
+pub trait SocketExt {
+    fn monitor(&self, monitor_endpoint: &str, events: i32) -> Result<()>;
+
+    fn is_ipv6(&self) -> Result<bool>;
+    fn set_ipv6(&self, value: bool) -> Result<()>;
```

Comment:
I'm assuming that setting any of these doesn't require a &mut?

Reply:
Indeed, it only requires &self. Later I want to extract some of this API to be a bit nicer, right now this was just a hack to test if exposing the methods will work.

Reply:
You could possibly just impl `Deref` on this to convert it to a `zmq::Message` if you don't need it to be mut, saves having to keep the methods up to date.

Reply:
That was my first idea, but then we would expose API that I would like to keep hidden. For example the user could call `bind`/`connect` even after the original TMQ socket has already been created (and therefore bound/connected). There's actually nothing wrong with that, I tried it and it works (even with the `Future` stuff on top of it), but it can have weird consequences.

For example, if you have a `Pull` socket bound to address A, it receives two multiparts and you `bind` to address B after reading the first multipart, but before reading the second, you might still get the second multipart even after you are bound to B. This would probably happen even if you used the synchronous sockets, but I wanted to disallow `bind`/`connect` happening after the initial socket setup to avoid having to deal with the semantics.